File: visualizer/renderer.py
"""
Rendering engine for different visualization modes.
Fixed: proper background clearing to prevent burn-in.
"""

import pygame
import numpy as np
import math
from typing import Tuple

from .settings import VisualizerSettings, ColorTheme, VISUALIZATION_MODES
from .effects import ParticleSystem, GlowEffect, WaterfallBuffer


class Renderer:
    """Main rendering class for all visualization modes."""

    def __init__(self, settings: VisualizerSettings):
        self.settings = settings
        self.particle_system = ParticleSystem(settings.particle_count)
        self.glow = GlowEffect()
        self.waterfall: WaterfallBuffer = None
        self.trail_surface: pygame.Surface = None
        self._prev_bands = None
        self._history: list = []
        self._max_history = 200

    def _ensure_surfaces(self, surface: pygame.Surface):
        """Create or resize surfaces as needed."""
        w, h = surface.get_size()

        if self.trail_surface is None or self.trail_surface.get_size() != (w, h):
            self.trail_surface = pygame.Surface((w, h))
            self.trail_surface.fill(self.settings.get_theme().background)

        if self.waterfall is None or (self.waterfall.width, self.waterfall.height) != (w, h):
            self.waterfall = WaterfallBuffer(w, h)

    def render(self, surface: pygame.Surface, band_data: np.ndarray,
               waveform: np.ndarray, mode: int, theme: ColorTheme):
        """Render the current visualization mode."""
        self._ensure_surfaces(surface)
        w, h = surface.get_size()

        # ── Background clearing ──
        # Always do a solid fill first, then optionally composite the trail
        bg = theme.background

        if self.settings.fade_trail:
            # Darken trail surface toward background color each frame
            fade_surface = pygame.Surface((w, h), pygame.SRCALPHA)
            # Higher fade_speed = faster fade = less burn-in
            fade_alpha = max(15, int(self.settings.fade_speed * 3.5))
            fade_alpha = min(255, fade_alpha)
            fade_surface.fill((*bg, fade_alpha))
            self.trail_surface.blit(fade_surface, (0, 0))
            # Copy trail onto main surface
            surface.blit(self.trail_surface, (0, 0))
        else:
            # No trail: hard clear every frame
            surface.fill(bg)
            self.trail_surface.fill(bg)

        # Store history for waterfall
        self._history.append(band_data.copy())
        if len(self._history) > self._max_history:
            self._history.pop(0)

        mode_name = VISUALIZATION_MODES[mode] if mode < len(VISUALIZATION_MODES) else "Bars"

        # Choose render target: draw to trail_surface so it persists for fade,
        # then blit trail onto main surface
        draw_target = self.trail_surface if self.settings.fade_trail else surface

        if mode_name == "Bars":
            self._draw_bars(draw_target, band_data, theme, w, h, mirrored=False)
        elif mode_name == "Bars Mirrored":
            self._draw_bars(draw_target, band_data, theme, w, h, mirrored=True)
        elif mode_name == "Circular":
            self._draw_circular(draw_target, band_data, theme, w, h)
        elif mode_name == "Wave":
            self._draw_wave(draw_target, waveform, band_data, theme, w, h)
        elif mode_name == "Spectrum Line":
            self._draw_spectrum_line(draw_target, band_data, theme, w, h)
        elif mode_name == "Particles":
            self._draw_particle_mode(surface, band_data, theme, w, h)
        elif mode_name == "Waterfall":
            self._draw_waterfall(surface, band_data, theme, w, h)

        # If we drew on trail_surface, composite it onto main surface
        if self.settings.fade_trail and mode_name not in ("Particles", "Waterfall"):
            surface.blit(self.trail_surface, (0, 0))

        # Draw particles overlay if enabled
        if self.settings.particles_enabled and mode_name != "Particles":
            self._emit_beat_particles(band_data, theme, w, h)

        dt = 1.0 / max(1, self.settings.fps_cap)
        self.particle_system.update(dt)
        self.particle_system.draw(surface)

        self._prev_bands = band_data.copy()

    def _lerp_color(self, c1: Tuple, c2: Tuple, t: float) -> Tuple[int, int, int]:
        """Linearly interpolate between two colors."""
        t = max(0.0, min(1.0, t))
        return (
            int(c1[0] + (c2[0] - c1[0]) * t),
            int(c1[1] + (c2[1] - c1[1]) * t),
            int(c1[2] + (c2[2] - c1[2]) * t),
        )

    def _get_bar_color(self, magnitude: float, index: int, total: int,
                       theme: ColorTheme) -> Tuple[int, int, int]:
        """Get color for a bar based on its position and magnitude."""
        position = index / max(1, total - 1)

        if position < 0.5:
            base_color = self._lerp_color(theme.primary, theme.secondary, position * 2)
        else:
            base_color = self._lerp_color(theme.secondary, theme.tertiary, (position - 0.5) * 2)

        # Brighten with magnitude
        brightness = 0.4 + 0.6 * magnitude
        return tuple(max(0, min(255, int(c * brightness))) for c in base_color)

    def _draw_bars(self, surface: pygame.Surface, data: np.ndarray,
                   theme: ColorTheme, w: int, h: int, mirrored: bool):
        """Draw bar visualization."""
        num_bars = len(data)
        total_width = w * 0.9
        bar_gap = total_width / num_bars
        bar_width = max(2, int(bar_gap * self.settings.bar_width_ratio))
        start_x = (w - total_width) / 2
        max_height = h * 0.85

        if mirrored:
            max_height = h * 0.42
            center_y = h // 2
        else:
            center_y = h

        for i, magnitude in enumerate(data):
            x = int(start_x + i * bar_gap)
            bar_height = int(magnitude * max_height)
            bar_height = max(2, bar_height)

            color = self._get_bar_color(magnitude, i, num_bars, theme)

            if mirrored:
                rect_up = pygame.Rect(x, center_y - bar_height, bar_width, bar_height)
                pygame.draw.rect(surface, color, rect_up,
                                 border_radius=self.settings.border_radius)
                rect_down = pygame.Rect(x, center_y, bar_width, bar_height)
                pygame.draw.rect(surface, color, rect_down,
                                 border_radius=self.settings.border_radius)

                if self.settings.glow and magnitude > 0.3:
                    self.glow.draw_glow_rect(surface, color, rect_up, magnitude * 0.5)
                    self.glow.draw_glow_rect(surface, color, rect_down, magnitude * 0.5)
            else:
                rect = pygame.Rect(x, center_y - bar_height, bar_width, bar_height)
                pygame.draw.rect(surface, color, rect,
                                 border_radius=self.settings.border_radius)

                if self.settings.glow and magnitude > 0.3:
                    self.glow.draw_glow_rect(surface, color, rect, magnitude * 0.5)

    def _draw_circular(self, surface: pygame.Surface, data: np.ndarray,
                       theme: ColorTheme, w: int, h: int):
        """Draw circular visualization."""
        cx, cy = w // 2, h // 2
        base_radius = min(w, h) * 0.15
        max_bar_length = min(w, h) * 0.3
        num_bars = len(data)

        for i, magnitude in enumerate(data):
            angle = (2 * math.pi * i / num_bars) - math.pi / 2
            bar_length = magnitude * max_bar_length + 5

            x1 = cx + math.cos(angle) * base_radius
            y1 = cy + math.sin(angle) * base_radius
            x2 = cx + math.cos(angle) * (base_radius + bar_length)
            y2 = cy + math.sin(angle) * (base_radius + bar_length)

            color = self._get_bar_color(magnitude, i, num_bars, theme)
            thickness = max(2, int(3 * (1 + magnitude)))

            pygame.draw.line(surface, color, (int(x1), int(y1)),
                             (int(x2), int(y2)), thickness)

            if self.settings.glow and magnitude > 0.5:
                self.glow.draw_glow_circle(surface, color,
                                           (int(x2), int(y2)), 4, magnitude * 0.4)

        pygame.draw.circle(surface, theme.primary, (cx, cy), int(base_radius), 2)

        avg_magnitude = float(np.mean(data))
        inner_radius = int(base_radius * 0.8 * (0.8 + 0.2 * avg_magnitude))
        inner_color = self._lerp_color(theme.background, theme.primary, avg_magnitude * 0.5)
        pygame.draw.circle(surface, inner_color, (cx, cy), inner_radius)

    def _draw_wave(self, surface: pygame.Surface, waveform: np.ndarray,
                   band_data: np.ndarray, theme: ColorTheme, w: int, h: int):
        """Draw waveform visualization."""
        center_y = h // 2
        amplitude = h * 0.35

        step = max(1, len(waveform) // w)
        samples = waveform[::step][:w]

        for layer in range(3):
            points = []
            color_t = layer / 3.0
            color = self._lerp_color(theme.primary, theme.tertiary, color_t)
            alpha_color = tuple(max(0, min(255, int(c * (1.0 - layer * 0.25)))) for c in color)

            for i, sample in enumerate(samples):
                x = i
                y = center_y + int(sample * amplitude * (1 + layer * 0.3))
                points.append((x, y))

            if len(points) > 2:
                pygame.draw.lines(surface, alpha_color, False, points,
                                  max(1, 3 - layer))

        avg_energy = float(np.mean(band_data))
        if self.settings.glow and avg_energy > 0.2:
            self.glow.draw_glow_circle(
                surface, theme.primary,
                (w // 2, center_y), int(20 * avg_energy), avg_energy * 0.3
            )

    def _draw_spectrum_line(self, surface: pygame.Surface, data: np.ndarray,
                            theme: ColorTheme, w: int, h: int):
        """Draw a smooth spectrum line with fill."""
        margin = w * 0.05
        usable_width = w - 2 * margin
        num_points = len(data)

        points_top = []
        points_bottom = []
        center_y = h * 0.6

        for i, magnitude in enumerate(data):
            x = int(margin + (i / max(1, num_points - 1)) * usable_width)
            y_offset = magnitude * h * 0.5
            points_top.append((x, int(center_y - y_offset)))
            points_bottom.append((x, int(center_y + y_offset * 0.3)))

        if len(points_top) > 2:
            fill_points = points_top + list(reversed(points_bottom))
            fill_surface = pygame.Surface((w, h), pygame.SRCALPHA)
            pygame.draw.polygon(fill_surface,
                                (*theme.primary, 40), fill_points)
            surface.blit(fill_surface, (0, 0))

            pygame.draw.lines(surface, theme.primary, False, points_top, 2)
            pygame.draw.lines(surface, theme.secondary, False, points_bottom, 1)

            for i, (x, y) in enumerate(points_top):
                if data[i] > 0.4:
                    color = self._get_bar_color(data[i], i, num_points, theme)
                    pygame.draw.circle(surface, color, (x, y), 3)
                    if self.settings.glow and data[i] > 0.6:
                        self.glow.draw_glow_circle(surface, color, (x, y), 5,
                                                   data[i] * 0.4)

    def _draw_particle_mode(self, surface: pygame.Surface, data: np.ndarray,
                            theme: ColorTheme, w: int, h: int):
        """Particle-based visualization."""
        num_bands = len(data)
        band_width = w / num_bands

        for i, magnitude in enumerate(data):
            if magnitude > 0.15:
                x = int((i + 0.5) * band_width)
                count = max(1, int(magnitude * 5))
                color = self._get_bar_color(magnitude, i, num_bands, theme)

                self.particle_system.emit(
                    x, h - 10, color,
                    count=count,
                    spread=band_width * 0.3,
                    speed=magnitude * 8
                )

    def _draw_waterfall(self, surface: pygame.Surface, data: np.ndarray,
                        theme: ColorTheme, w: int, h: int):
        """Draw waterfall/spectrogram visualization."""
        self.waterfall.update(data, theme)
        self.waterfall.draw(surface)

    def _emit_beat_particles(self, data: np.ndarray, theme: ColorTheme,
                             w: int, h: int):
        """Emit particles on beats (energy spikes)."""
        if self._prev_bands is None:
            return

        diff = data - self._prev_bands
        for i, d in enumerate(diff):
            if d > 0.3:
                x = int((i / len(data)) * w)
                color = self._get_bar_color(data[i], i, len(data), theme)
                self.particle_system.emit(x, h * 0.5, color, count=2, speed=4)

File: src/renderer.c
#include "renderer.h"
#include <math.h>
#include <stdlib.h>
#include <string.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

static void set_color(SDL_Renderer *r, Color c, uint8_t a) {
    SDL_SetRenderDrawColor(r, c.r, c.g, c.b, a);
}

static Color get_bar_color(float magnitude, int index, int total, ColorTheme *theme) {
    float position = (total > 1) ? (float)index / (float)(total - 1) : 0.0f;
    Color base;
    if (position < 0.5f)
        base = color_lerp(theme->primary, theme->secondary, position * 2.0f);
    else
        base = color_lerp(theme->secondary, theme->tertiary, (position - 0.5f) * 2.0f);

    float brightness = 0.4f + 0.6f * magnitude;
    return color_brightness(base, brightness);
}

/* ── Particles ─────────────────────────────────────────────── */

static void particle_emit(Renderer *r, float x, float y, Color color,
                          int count, float spread, float speed) {
    for (int i = 0; i < count && r->particle_count < MAX_PARTICLES; i++) {
        Particle *p = &r->particles[r->particle_count++];
        p->x = x;
        p->y = y;
        p->vx = ((float)rand() / RAND_MAX - 0.5f) * spread * 2.0f;
        p->vy = -((float)rand() / RAND_MAX * speed * 1.5f + speed * 0.5f);
        p->lifetime = 0.5f + (float)rand() / RAND_MAX * 1.5f;
        p->max_lifetime = p->lifetime;
        p->size = 1.5f + (float)rand() / RAND_MAX * 2.5f;
        p->color = color;
        p->alive = true;
    }
}

static void particles_update(Renderer *r, float dt) {
    int alive = 0;
    for (int i = 0; i < r->particle_count; i++) {
        Particle *p = &r->particles[i];
        if (!p->alive) continue;

        p->x += p->vx * dt * 60.0f;
        p->y += p->vy * dt * 60.0f;
        p->vy += 0.05f * dt * 60.0f;
        p->lifetime -= dt;

        if (p->lifetime <= 0.0f) {
            p->alive = false;
        } else {
            r->particles[alive++] = *p;
        }
    }
    r->particle_count = alive;
}

static void particles_draw(Renderer *r, SDL_Renderer *sdl) {
    for (int i = 0; i < r->particle_count; i++) {
        Particle *p = &r->particles[i];
        if (!p->alive) continue;

        float alpha_f = p->lifetime / p->max_lifetime;
        uint8_t a = (uint8_t)(alpha_f * 255.0f);
        int size = (int)(p->size * alpha_f);
        if (size < 1) size = 1;

        Color c = color_brightness(p->color, alpha_f);
        SDL_SetRenderDrawColor(sdl, c.r, c.g, c.b, a);

        SDL_Rect rect = {(int)p->x - size / 2, (int)p->y - size / 2, size, size};
        SDL_RenderFillRect(sdl, &rect);
    }
}

/* ── Visualization modes ───────────────────────────────────── */

static void draw_bars(SDL_Renderer *sdl, float *data, int num_bars,
                      ColorTheme *theme, int w, int h, bool mirrored,
                      float bar_width_ratio) {
    float total_width = w * 0.9f;
    float bar_gap = total_width / num_bars;
    int bar_width = (int)(bar_gap * bar_width_ratio);
    if (bar_width < 2) bar_width = 2;
    float start_x = (w - total_width) / 2.0f;

    float max_height = mirrored ? h * 0.42f : h * 0.85f;
    int center_y = mirrored ? h / 2 : h;

    for (int i = 0; i < num_bars; i++) {
        int x = (int)(start_x + i * bar_gap);
        int bar_height = (int)(data[i] * max_height);
        if (bar_height < 2) bar_height = 2;

        Color c = get_bar_color(data[i], i, num_bars, theme);

        if (mirrored) {
            SDL_Rect up = {x, center_y - bar_height, bar_width, bar_height};
            SDL_Rect down = {x, center_y, bar_width, bar_height};
            set_color(sdl, c, 255);
            SDL_RenderFillRect(sdl, &up);
            SDL_RenderFillRect(sdl, &down);
        } else {
            SDL_Rect rect = {x, center_y - bar_height, bar_width, bar_height};
            set_color(sdl, c, 255);
            SDL_RenderFillRect(sdl, &rect);
        }
    }
}

static void draw_circular(SDL_Renderer *sdl, float *data, int num_bars,
                           ColorTheme *theme, int w, int h) {
    int cx = w / 2, cy = h / 2;
    float base_r = (w < h ? w : h) * 0.15f;
    float max_len = (w < h ? w : h) * 0.3f;

    for (int i = 0; i < num_bars; i++) {
        float angle = (2.0f * (float)M_PI * i / num_bars) - (float)M_PI / 2.0f;
        float bar_len = data[i] * max_len + 5.0f;

        int x1 = cx + (int)(cosf(angle) * base_r);
        int y1 = cy + (int)(sinf(angle) * base_r);
        int x2 = cx + (int)(cosf(angle) * (base_r + bar_len));
        int y2 = cy + (int)(sinf(angle) * (base_r + bar_len));

        Color c = get_bar_color(data[i], i, num_bars, theme);
        set_color(sdl, c, 255);
        SDL_RenderDrawLine(sdl, x1, y1, x2, y2);
    }

    /* Center circle outline */
    set_color(sdl, theme->primary, 200);
    int segments = 60;
    for (int i = 0; i < segments; i++) {
        float a1 = 2.0f * (float)M_PI * i / segments;
        float a2 = 2.0f * (float)M_PI * (i + 1) / segments;
        SDL_RenderDrawLine(sdl,
            cx + (int)(cosf(a1) * base_r), cy + (int)(sinf(a1) * base_r),
            cx + (int)(cosf(a2) * base_r), cy + (int)(sinf(a2) * base_r));
    }
}

static void draw_wave(SDL_Renderer *sdl, float *waveform, int waveform_size,
                      ColorTheme *theme, int w, int h) {
    int center_y = h / 2;
    float amplitude = h * 0.35f;

    int step = waveform_size / w;
    if (step < 1) step = 1;

    Color colors[3];
    colors[0] = theme->primary;
    colors[1] = color_lerp(theme->primary, theme->tertiary, 0.33f);
    colors[2] = color_lerp(theme->primary, theme->tertiary, 0.66f);

    for (int layer = 0; layer < 3; layer++) {
        Color c = color_brightness(colors[layer], 1.0f - layer * 0.25f);
        set_color(sdl, c, 255);

        int prev_x = 0;
        int prev_y = center_y;

        for (int x = 0; x < w; x++) {
            int sample_idx = x * step;
            if (sample_idx >= waveform_size) sample_idx = waveform_size - 1;

            float sample = waveform[sample_idx];
            int y = center_y + (int)(sample * amplitude * (1.0f + layer * 0.3f));

            if (x > 0) {
                SDL_RenderDrawLine(sdl, prev_x, prev_y, x, y);
            }
            prev_x = x;
            prev_y = y;
        }
    }
}

static void draw_spectrum_line(SDL_Renderer *sdl, float *data, int num_points,
                               ColorTheme *theme, int w, int h) {
    float margin = w * 0.05f;
    float usable = w - 2.0f * margin;
    int center_y = (int)(h * 0.6f);

    /* Top line */
    set_color(sdl, theme->primary, 255);
    int prev_x = (int)margin, prev_y = center_y;

    for (int i = 0; i < num_points; i++) {
        int x = (int)(margin + ((float)i / (num_points - 1)) * usable);
        int y = center_y - (int)(data[i] * h * 0.5f);

        if (i > 0) SDL_RenderDrawLine(sdl, prev_x, prev_y, x, y);
        prev_x = x;
        prev_y = y;

        /* Dots at peaks */
        if (data[i] > 0.4f) {
            Color c = get_bar_color(data[i], i, num_points, theme);
            set_color(sdl, c, 255);
            SDL_Rect dot = {x - 2, y - 2, 5, 5};
            SDL_RenderFillRect(sdl, &dot);
            set_color(sdl, theme->primary, 255);
        }
    }

    /* Bottom line */
    set_color(sdl, theme->secondary, 180);
    prev_x = (int)margin;
    prev_y = center_y;

    for (int i = 0; i < num_points; i++) {
        int x = (int)(margin + ((float)i / (num_points - 1)) * usable);
        int y = center_y + (int)(data[i] * h * 0.15f);
        if (i > 0) SDL_RenderDrawLine(sdl, prev_x, prev_y, x, y);
        prev_x = x;
        prev_y = y;
    }
}

static void draw_particle_mode(Renderer *r, SDL_Renderer *sdl, float *data,
                                int num_bands, ColorTheme *theme, int w, int h) {
    (void)sdl;  /* ← ADD THIS LINE */
    float band_width = (float)w / num_bands;

    for (int i = 0; i < num_bands; i++) {
        if (data[i] > 0.15f) {
            float x = (i + 0.5f) * band_width;
            int count = (int)(data[i] * 5);
            if (count < 1) count = 1;
            Color c = get_bar_color(data[i], i, num_bands, theme);
            particle_emit(r, x, (float)(h - 10), c, count,
                          band_width * 0.3f, data[i] * 8.0f);
        }
    }
}

static void draw_waterfall(Renderer *r, SDL_Renderer *sdl, float *data,
                           int num_bands, ColorTheme *theme, int w, int h) {
    /* Ensure texture exists */
    if (!r->waterfall_texture || r->waterfall_w != w || r->waterfall_h != h) {
        if (r->waterfall_texture) SDL_DestroyTexture(r->waterfall_texture);
        r->waterfall_texture = SDL_CreateTexture(sdl, SDL_PIXELFORMAT_RGBA8888,
            SDL_TEXTUREACCESS_TARGET, w, h);
        r->waterfall_w = w;
        r->waterfall_h = h;

        /* Clear */
        SDL_SetRenderTarget(sdl, r->waterfall_texture);
        SDL_SetRenderDrawColor(sdl, 0, 0, 0, 255);
        SDL_RenderClear(sdl);
        SDL_SetRenderTarget(sdl, NULL);
    }

    /* Scroll up: copy texture shifted */
    SDL_Texture *temp = SDL_CreateTexture(sdl, SDL_PIXELFORMAT_RGBA8888,
        SDL_TEXTUREACCESS_TARGET, w, h);
    SDL_SetRenderTarget(sdl, temp);
    SDL_SetRenderDrawColor(sdl, 0, 0, 0, 255);
    SDL_RenderClear(sdl);

    SDL_Rect src = {0, 1, w, h - 1};
    SDL_Rect dst = {0, 0, w, h - 1};
    SDL_RenderCopy(sdl, r->waterfall_texture, &src, &dst);

    /* Draw new line at bottom */
    float band_w = (float)w / num_bands;
    for (int i = 0; i < num_bands; i++) {
        float mag = data[i];
        Color c = color_lerp(theme->secondary, theme->primary, mag);
        float brightness = mag * 2.0f;
        if (brightness > 1.0f) brightness = 1.0f;
        c = color_brightness(c, brightness);

        SDL_SetRenderDrawColor(sdl, c.r, c.g, c.b, 255);
        SDL_Rect pixel = {(int)(i * band_w), h - 1, (int)(band_w) + 1, 1};
        SDL_RenderFillRect(sdl, &pixel);
    }

    SDL_SetRenderTarget(sdl, NULL);

    /* Swap */
    SDL_DestroyTexture(r->waterfall_texture);
    r->waterfall_texture = temp;

    /* Draw to screen */
    SDL_RenderCopy(sdl, r->waterfall_texture, NULL, NULL);
}

/* ── Main renderer functions ───────────────────────────────── */

void renderer_init(Renderer *r, Settings *settings) {
    memset(r, 0, sizeof(Renderer));
    r->settings = settings;
}

void renderer_destroy(Renderer *r, SDL_Renderer *sdl_renderer) {
    (void)sdl_renderer;
    if (r->trail_texture) SDL_DestroyTexture(r->trail_texture);
    if (r->waterfall_texture) SDL_DestroyTexture(r->waterfall_texture);
    memset(r, 0, sizeof(Renderer));
}

void renderer_render(Renderer *r, SDL_Renderer *sdl,
                     float *band_data, int num_bands,
                     float *waveform, int waveform_size,
                     int mode, ColorTheme *theme) {
    int w, h;
    SDL_GetRendererOutputSize(sdl, &w, &h);

    /* Clear background */
    Color bg = theme->background;

    if (r->settings->fade_trail && mode != 5 && mode != 6) {
        /* Fade effect: draw semi-transparent background */
        int alpha = (int)(r->settings->fade_speed * 3.5f);
        if (alpha < 15) alpha = 15;
        if (alpha > 255) alpha = 255;

        SDL_SetRenderDrawBlendMode(sdl, SDL_BLENDMODE_BLEND);
        SDL_SetRenderDrawColor(sdl, bg.r, bg.g, bg.b, (uint8_t)alpha);
        SDL_Rect full = {0, 0, w, h};
        SDL_RenderFillRect(sdl, &full);
    } else {
        SDL_SetRenderDrawColor(sdl, bg.r, bg.g, bg.b, 255);
        SDL_RenderClear(sdl);
    }

    /* Render current mode */
    switch (mode) {
        case 0: /* Bars */
            draw_bars(sdl, band_data, num_bands, theme, w, h, false,
                      r->settings->bar_width_ratio);
            break;
        case 1: /* Bars Mirrored */
            draw_bars(sdl, band_data, num_bands, theme, w, h, true,
                      r->settings->bar_width_ratio);
            break;
        case 2: /* Circular */
            draw_circular(sdl, band_data, num_bands, theme, w, h);
            break;
        case 3: /* Wave */
            draw_wave(sdl, waveform, waveform_size, theme, w, h);
            break;
        case 4: /* Spectrum Line */
            draw_spectrum_line(sdl, band_data, num_bands, theme, w, h);
            break;
        case 5: /* Particles */
            SDL_SetRenderDrawColor(sdl, bg.r, bg.g, bg.b, 255);
            SDL_RenderClear(sdl);
            draw_particle_mode(r, sdl, band_data, num_bands, theme, w, h);
            break;
        case 6: /* Waterfall */
            draw_waterfall(r, sdl, band_data, num_bands, theme, w, h);
            break;
    }

    /* Beat-reactive particles */
    if (r->settings->particles_enabled && mode != 5) {
        if (r->has_prev) {
            for (int i = 0; i < num_bands; i++) {
                float diff = band_data[i] - r->prev_bands[i];
                if (diff > 0.3f) {
                    float x = ((float)i / num_bands) * w;
                    Color c = get_bar_color(band_data[i], i, num_bands, theme);
                    particle_emit(r, x, h * 0.5f, c, 2, 3.0f, 4.0f);
                }
            }
        }
    }

    /* Update and draw particles */
    float dt = 1.0f / (r->settings->fps_cap > 0 ? r->settings->fps_cap : 60);
    particles_update(r, dt);
    SDL_SetRenderDrawBlendMode(sdl, SDL_BLENDMODE_BLEND);
    particles_draw(r, sdl);

    /* Store previous bands */
    memcpy(r->prev_bands, band_data, sizeof(float) * num_bands);
    r->has_prev = true;
}